client/src/common/fileOperations.ts

```diff
@@ -43,17 +60,20 @@ abstract class FileOperationFeature<I, E extends { readonly files: ReadonlyArray
 	private _serverCapability: keyof proto.FileOperationOptions;
 	private _listener: code.Disposable | undefined;
 	private _globPatterns: Map<string, RegExp> = new Map<string, RegExp>();
+	protected fixSlashes: (e: E) => Promise<E>;
 
 	constructor(client: BaseLanguageClient, event: code.Event<E>,
 		registrationType: proto.RegistrationType<proto.FileOperationRegistrationOptions>,
 		clientCapability: keyof proto.FileOperationClientCapabilities,
-		serverCapability: keyof proto.FileOperationOptions)
+		serverCapability: keyof proto.FileOperationOptions,
+		fixSlashes: (e: E) => Promise<E>)
```

Comment:
I don't think we need to make this a instance field. It is always the same function we use.

Reply:
It's slightly different for renames:

```ts
async (params) => ({ files: await Promise.all(params.files.map(addSlashesToDirectories)) })
// vs
async (params) => ({ files: await Promise.all(params.files.map(async (i) => ({ oldUri: await addSlashesToDirectories(i.oldUri), newUri: await addSlashesToDirectories(i.newUri) }))) })
```

I think I have a solution to my async stuff (including an easy way to remove the `waitUntil` from the middleware so it can't try to use it - since it would very likely break things, and it should just be returning the `Promise<WorkspaceEdit>`).

Just doing some more testing and I'll push it up for you to review.

Reply:
What I mean is that for example in a rename we only need to transform the oldUri since this is the URI we test again. The new URI shouldn't be transformed. In addition we shouldn't temper with the URI itself. The '/' shouldn't reach the server since URI for folders don't reach the server in that format in all other cases.

Reply:
Ah, got it! I'll fix :-)

Reply:
What we can do if necessary is to add a hint to the LSP params whether the URI represents a directory or file.